### ISSUE TITLE
Custom components: passing transitions

### ISSUE THREAD

**Issue by paulpflug** [vuejs/vue#3678]
I'm currently in the process to make my custom components flexible in terms of transitions.
So this is what I want:
```html
<some-custom-component transition="fade"></some-custom-component>
```
(the custom component has the logic where and when the transition should be applied)

There are some problems I encountered (v1.0.0)
- Vue gives a warning "Transitions will not work on a fragment instance.". When the component is a fragment instance and despite "transition" is a defined prop of that instance. Behaves as expected, only the warning is annoying.
- transitions are not resolved on the parent-scope
```coffee
template: "<some-custom-component transition='fade'></some-custom-component>"
components:
  someCustomComponent: require("some-custom-component")
transitions:
  fade: #will not be found
```
- I need to know when the transitions end or are canceled. 

From the current documentation of v2.0.0 - this seams nearly impossible in v2.0.0.
Just imagine a custom component with multiple different transitions - for example one transition for the fore- and one for the background.

Maybe it would be a good Idea to let transitions behave a bit like mixins:
```coffee
transitionMixins:
  fade:
    css: false
    enter: -> #fadeIn
    leave: -> #fadeOut
template: "<transition mixin='fade'><div></div></transition>
```

then this could be possible:
```coffee
template: "<some-custom-component transition1='fade' transition2='bounce'></some-custom-component>"
components:
  someCustomComponent: require("some-custom-component")
transitionMixins:
  fade: #fading transition
  bounce: #bouncing transition
```

inside `some-custom-component`
```coffee
template: "
  <div>
    <transition @afterLeave=finished v-bind:mixin=transition1><foreground/></transition>
    <transition v-bind:mixin=transition2><background/></transition>
  </div>
"
props: ["transition1","transition2"]
methods:
  finished: -> # called when transition1 is finished
```

I'm open for other methods to handle transition with custom components..

**Comment by fnlctrl** [comment]
Hi, thanks for filling this issue.
For 1.0, `transition` is a special reserved attribute which you shouldn't be using as a prop name.
Instead of passing a `transition` object, you can simply pass a transition name to a component, and use it inside its template.
``` html
<div :transition="dynamicTransitonName"></div>
```
But that only works with css-only transitions. For javascript transitions, in 1.0 you must define it as a global asset (Vue.transition(....))

Side note: for 1.0 since `transition` is a component asset type, you can indeed define it in a mixin and use it inside components
``` js
Vue.mixin('foo', {
  transitions: {
    yourTransition: {....}
  }
})
```

In 2.0, transition is no longer an attribute, and has been redesigned as a component. This makes things simpler as you can just pass the transition component to the child component
`<child :transition-component="$options.components.someTransitionComponent">`

**Comment by paulpflug** [comment]
First for 1.0:
My problem is, I need js hooks and I need access to these hooks on a user generated transition.
So I can order the user to do this:
```coffee
transitions:
  fade:
    enter: (el,done) ->
      this.$emit("before-enter") # will fire on the component scope - so I have access again
      # fading
      this.$emit("entered") 
      done()
```
But this is very difficult to debug for the user when he doesn't comply - because I can't check if he did. 


For 2.0
`<child :transition-component="$options.components.someTransitionComponent"/>`
How would I use this transition in the template?
```html
<transition :is=transition-component></transition>
```
is this possible:
```html
<transition :is=transition-component @after-enter=entered></transition>
```
will both, the `someTransitionComponent` hook and `entered` get called? in which order?
if, this would meet my needs.

Thanks for your time!

**Comment by fnlctrl** [comment]
For 1.0:
If I understand you correctly, you are trying to access a user-defined component's transition asset, which can be done by v-ref: `<child v-ref:child/>`, `this.$refs.child.$options.transitions.someTransitionName`

For 2.0:
`<child :transition-component="$options.components.someTransitionComponent"/>`
Inside child template:
`<div>
  <component :is="transitionComponent">
      <div>The transition is applied on this node</div>
  </component>
</div>
`
Note that child's `transitionComponent` prop's type is Function.
You can of course give it other names, as it's just a prop.
This is called passing parent's component asset to child, via props.

**Comment by paulpflug** [comment]
Thanks to you, I solved it for 1.0 with using `Vue.util.resolveAsset(this.$parent.$options,'transitions',this.transition)` to get the right transition hook.
Based on this I create a new transition with the hooks inserted that I need.
(I made a mixin out of it: [vue-mixins/transition](https://github.com/paulpflug/vue-mixins/blob/master/src/transition.coffee))

2.0 isn't important for me, yet. When I run into problems I will create a new issue.

Thanks again @fnlctrl :+1:
